### PR TITLE
Prototype extended-name-section proposal

### PR DESCRIPTION
**Pull request by dcodeIO**
Implements the parts of the [Extended Name Section Proposal for WebAssembly](https://github.com/WebAssembly/extended-name-section) that are trivially applicable to Binaryen, in particular:

* Table name<del>s</del>
* Memory name<del>s</del>
* Global names

Does not yet implement

* Label names
* Type names
* Elem names
* Data names

Given that unknown name section subsections will be ignored anyway, I didn't add a feature flag, but can add one if preferred.

Background: @binji [pointed out](https://github.com/WebAssembly/design/issues/1377#issuecomment-694424414) that there is already an existing proposal when I asked about support for global names as a follow-up to https://github.com/WebAssembly/binaryen/pull/3115#discussion_r486748807, and that

> to advance we really need only tools + engine support

As such, this PR, in case it is accepted, adds another tool to the list, if there is one already, with partial support. Will already be useful for us at AssemblyScript since we are using Binaryen outputs exclusively in testing, so we will have proper names, and I hope that more tools and engines will follow.

### CODE REVIEW

**Comment by dcodeIO** [comment]
Same CI issue as in https://github.com/WebAssembly/binaryen/pull/3154: `TypeError: y.get(...) is not a function`

**Comment by dcodeIO** [comment]
@kripken Is this a proposal Binaryen would be willing to support at this point in time, given that Extended Name Section is only in Phase 1 but needs tooling support to advance?

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -2280,8 +2339,65 @@ void WasmBinaryBuilder::readNames(size_t payloadLen) {
...
+                    << std::to_string(index) << std::endl;
+        }
+      }
+    } else if (nameType == BinaryConsts::UserSections::Subsection::NameMemory) {
+      auto num = getU32LEB();
+      for (size_t i = 0; i < num; i++) {
+        auto index = getU32LEB();
+        auto rawName = getInlineString();
+        rawName = escape(rawName);
+        auto name = rawName;
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:2364]: Is this supposed to look like this?

```suggestion
        auto name = escape(rawName);
```

Could also do `auto name = escape(getInlineString());`.

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -2280,8 +2339,65 @@ void WasmBinaryBuilder::readNames(size_t payloadLen) {
...
+        }
+      }
+    } else if (nameType == BinaryConsts::UserSections::Subsection::NameGlobal) {
+      auto num = getU32LEB();
+      std::set<Name> usedNames;
+      for (size_t i = 0; i < num; i++) {
+        auto index = getU32LEB();
+        auto rawName = getInlineString();
+        rawName = escape(rawName);
+        auto name = rawName;
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:2381]: Same variable pattern here.

File: src/wasm/wasm-binary.cpp [hunk 3]
```diff
@@ -2941,15 +3057,37 @@ void WasmBinaryBuilder::visitLocalSet(LocalSet* curr, uint8_t code) {
...
+    curr->name = import->name;
+    curr->type = import->type;
+  } else {
+    Index adjustedIndex = index - globalImports.size();
+    if (adjustedIndex >= globals.size()) {
+      throwError("invalid global index");
+    }
+    auto* glo = globals[adjustedIndex];
+    curr->name = glo->name;
+    curr->type = glo->type;
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:3071]: `glo` -> `glob` for consistency?

File: test/unit/input/update.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+# This file updates the target_feature tests in the rare but unfortunate case
+# that their roundtripped binary representations do not match the input anymore
+# due to otherwise unrelated binary format changes.
```
Comment (tlively) [inline test/unit/input/update.sh:5]: Nice 👍

**Review by tlively — APPROVED** [review]
Landing this SGTM. I don't think we need to complicate things with a feature flag as long as browsers still use the known subsections in the presence of the new unknown subsections.

File: src/wasm/wasm-binary.cpp [hunk 5]
```diff
@@ -1819,11 +1862,12 @@ void WasmBinaryBuilder::readGlobals() {
       throwError("Global mutability must be 0 or 1");
     }
     auto* init = readExpression();
-    wasm.addGlobal(
-      Builder::makeGlobal("global$" + std::to_string(i),
```
[the remark this answers is not in the document]
Comment (dcodeIO) [reply src/wasm/wasm-binary.cpp:1823]: Oh, here I mindlessly looked up what's done for functions that are not immediately added to the module but kept in a temporary list, and just copied that. Agree that using the builder here is better.

File: src/wasm/wasm-binary.cpp (hunk 1 above)
Comment (dcodeIO) [reply src/wasm/wasm-binary.cpp:2364]: This is an exact copy & paste from what was already done when reading functions

https://github.com/WebAssembly/binaryen/blob/9de15862c1bc4e2b092ccbe89fdb4b850704bbfc/src/wasm/wasm-binary.cpp#L2228-L2230

Going to update it there as well :)

File: test/binaryen.js/debug-names.js.txt [hunk 6]
```diff
@@ -27,7 +27,19 @@
  (table $wor 0 0 funcref)
  (global $ld i32 (i32.const 0))
  (func $of (param $js i32)
-  (local $!#$%&'*+-./:<=>?@\^_`|~ f64)
+  (local $!#$%&'*+-./:<=>?@\5c^_`|~ f64)
```
Comment (dcodeIO) [inline test/binaryen.js/debug-names.js.txt:30]: The replacement of `\` to `\5c` here is interesting in a way, in that it isn't strictly necessary to escape names in the debug section and `\` is a valid idchar as of the spec. Ideally, roundtripping a binary should not change this, and escaping is only really necessary when emitting text  format. Perhaps something to look into in a follow-up, but added a second roundtrip here to validate that names are read back correctly once escaped, i.e. that there's no double escaping.

**Comment by sbc100** [comment]
This change seems to have generated an extra table name subsection contain in the name "0" for the table for all emscripten binaries.

This adds 6 bytes to every binary, but more importantly I don't think its correct.. isn't "0" the internal name given to the table by binaryen?   llvm and emscripten certainly don't name the table.

**Comment by dcodeIO** [comment]
Yes, "0" is the default name given to both the memory and the table, but some test cases actually name it "0" explicitly in text format, so can't really be omitted in every case it seemed since these cannot be told apart. Name section changes are somewhat unfortunate for tests comparing exact binaries, like Binaryen's [unit tests](https://github.com/WebAssembly/binaryen/commit/fa4c884f4ebfde185c9d8a3ee4e54f96c57cebed#diff-f024eb4a40b154ea1a8ff3dd0beb0cf5), or otherwise expecting the section to never change, but https://github.com/emscripten-core/emscripten/pull/12375 seems like a good approach to deal with it.

**Comment by sbc100** [comment]
Surely we can find a way to distinguish between actual internal makes and internal names?

How does this for other things like memory or types?  Surely those things internal names too, but are not appearing in the name sections.  Why it is only the table? 

I think this is worth fixing, regardless of the sensitivity of the tests.  I don't binaryen should add names that were not present in the input file.  With this change it seems that tables appear to have a name if even if the input binary does not name its table.

**Comment by dcodeIO** [comment]
I believe at other places this is solved by calling a method like `getNameOrDefault`, returning either the name if it has been set or a default name that is typically a `Name::fromInt(0)` or something, in turn keeping the respective `name.isNull()` on the object, which can be detected when creating the debug section. Agree that it would be best to introduce a similar mechanism for memories and tables as well.

**Comment by sbc100** [comment]
Memories don't seem to have this problem by the way... its only tables.

**Comment by dcodeIO** [comment]
That's indeed strange, and something I'd have to investigate. The respective code part is here:

https://github.com/WebAssembly/binaryen/blob/7e99538386e047bddc41ad50b9762b7d4b9611b8/src/wasm/wasm-binary.cpp#L630-L646

And `name.is()` is used for both, indicating that if this is indeed the case that there's already some mechanism in place for memories?

**Pull request merged**